### PR TITLE
Dev 41 create comments section

### PR DESCRIPTION
**Pull request by olehrdk**
## Ticket
Link to ticket: https://bsalo.atlassian.net/jira/software/projects/DEV/boards/2/backlog?selectedIssue=DEV-41

## Description
*Please provide a small summary of your changes and how they resolved the issues in the ticket:*

- Comments Component added
- Tests for Comments component added
- BlogPost page added

## Screenshots

image

## Checklist

- [x] Did you test it?
- [x] Did you add tests?
- [x] Did you add to the storybook?

### CODE REVIEW

File: src/actions/BlogActions.tsx [hunk 1]
```diff
@@ -0,0 +1,12 @@
+enum BlogActionTypes {
+  COMMENT_ADDED = '[Blog]: Comment Added',
+}
+
+class BlogActions {
+  static addComment = (payload: { value: string; postId: number; parentId: number | null }) => ({
```
Comment (dev1B) [inline src/actions/BlogActions.tsx:6]: if u adding new comment u will not have postId

File: src/actions/BlogActions.tsx [hunk 2]
```diff
@@ -0,0 +1,12 @@
+enum BlogActionTypes {
+  COMMENT_ADDED = '[Blog]: Comment Added',
```
Comment (dev1B) [inline src/actions/BlogActions.tsx:2]: ADD_COMMENT

Comment (dev1B) [reply src/actions/BlogActions.tsx:2]: Add Comment

**Review by dev1B — CHANGES_REQUESTED** [review]

File: src/actions/BlogActions.tsx (hunk 1 above)
Comment (olehrdk) [reply src/actions/BlogActions.tsx:6]: postId needed only to find post where we have to add comment

File: src/components/Comments/Comments.test.tsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    commentText:
+      'Lorem ipsum dolor sit amet consectetur adipisicing elit. Laborum neque distinctio praesentium quisquam laudantium tempore? Illum fugit esse nihil obcaecati?',
+    userName: 'John Newman 3',
+    userAvatar: null,
+  },
+];
+describe('Comments', () => {
+  it('Should successfully render', () => {
+    const wrapper = renderer
+      .create(<Comments commentsList={commentsList} onCommentAdded={() => console.log('added')} />)
```
Comment (dev1B) [inline src/components/Comments/Comments.test.tsx:29]: jest.fn

File: src/components/AddCommentForm/AddCommentForm.tsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+interface AddCommentFormProps {
+  onSubmit: (value: string) => void;
+  focus: boolean;
+}
+const focusOnTextArea = (ref: any) => {
+  ref.current?.focus({ cursor: 'end' });
+};
+
+const AddCommentForm = ({ onSubmit, focus }: AddCommentFormProps) => {
+  const [textAreaValue, setTextAreaValue] = useState('');
```
Comment (dev1B) [inline src/components/AddCommentForm/AddCommentForm.tsx:15]: use form hooks please

**Review by dev1B — CHANGES_REQUESTED** [review]

File: src/components/AddCommentForm/AddCommentForm.tsx [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+         className="comments__form"
+         onFinish={() => {
+            onSubmit(textAreaValue);
+            setTextAreaValue('');
+         }}>
+         <Form.Item>
+            <TextArea
+               ref={textAreaRef}
+               value={textAreaValue}
+               onChange={(e) => { setTextAreaValue(e.target.value) }}
```
Comment (Skitus) [inline src/components/AddCommentForm/AddCommentForm.tsx]: Use react-hook-form. Don`t develop your logic(u don`t need to hang listener)

File: src/components/AddCommentForm/AddCommentForm.tsx [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+               rows={4}
+            />
+         </Form.Item>
+         <Form.Item>
+            <Button
+               disabled={disabled}
```
Comment (Skitus) [inline src/components/AddCommentForm/AddCommentForm.tsx]: use react-hook form. Why u use this button disabled ? Add some logic. For instance add your comment to console

File: src/pages/BlogPost/BlogPostContainer.tsx [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+   }
+}
+
+const mapDispatchToProps = (dispatch: typeof store.dispatch) => {
+   return {
+      onCommentAdded: (value: string, parentId: number | null) => dispatch(BlogActions.addComment({ value, postId, parentId }))
+   }
+}
+
+export default connect<StateProps, DispatchProps, OwnProps, InitialState>(mapStateToProps, mapDispatchToProps)(BlogPost);
```
Comment (Skitus) [inline src/pages/BlogPost/BlogPostContainer.tsx]: Are we have to use Class components ?

File: src/reducers/blog.tsx [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+   switch (action.type) {
+      case BlogActionTypes.COMMENT_ADDED:
+         const { postId, value, parentId } = action.payload;
+
+         const postIdx = state.posts.findIndex((post: Post) => post.postId === postId)
+         const { comments } = state.posts[postIdx]
+
+         const newComment: comment = {
+            commentId: comments.length + 1,
```
Comment (Skitus) [inline src/reducers/blog.tsx]: Mb better use comments.length++

File: src/reducers/blog.tsx [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            parentId: parentId,
+            children: [],
+            commentText: value,
+            userName: `Anonim ${comments.length + 1}`,
```
Comment (Skitus) [inline src/reducers/blog.tsx]: Mb better use comments.length++ or ++comments.length

File: src/components/AddCommentForm/AddCommentForm.tsx (hunk 6 above)
Comment (olehrdk) [reply src/components/AddCommentForm/AddCommentForm.tsx]: Button disabled when textarea value length = 0
Console.log throws an eslint error
Which logic I have to add?

File: src/pages/BlogPost/BlogPostContainer.tsx (hunk 7 above)
Comment (olehrdk) [reply src/pages/BlogPost/BlogPostContainer.tsx]: There aren't any Class Components

File: src/components/Comments/Comments.tsx [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+const Comments = ({ commentsList, onCommentAdded }: CommentsProps) => {
+  const [replyTo, setReplyTo] = useState(null);
+  let focus: boolean = false;
+  if (replyTo) focus = true;
+  let commentsListHTML = <p>No comments yet. Be the first</p>;
+  const nested = nestComments(commentsList);
+
+  if (commentsList && commentsList.length !== 0) {
```
Comment (dev1B) [inline src/components/Comments/Comments.tsx:20]: commentsList && commentsList.length !== 0 => commentsList?.length better)

File: src/utils/scrollToElement.tsx [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import { scroller } from 'react-scroll';
+
+function scrollToElement(className: string): void {
```
Comment (dev1B) [inline src/utils/scrollToElement.tsx:3]: scrollToElement => scrollTo

File: src/types/blogTypes.tsx [hunk 12]
```diff
@@ -0,0 +1,17 @@
+export interface comment {
```
Comment (dev1B) [inline src/types/blogTypes.tsx:1]: comment => Comment

File: src/types/blogTypes.tsx [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+  commentId: number;
+  parentId: number | null;
+  children: [];
```
Comment (dev1B) [inline src/types/blogTypes.tsx:4]: ?

File: src/types/blogTypes.tsx [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+  userName: string;
+  userAvatar?: string | null;
+}
+
+export interface Post {
+  postId: number;
+  comments: comment[];
+}
+
+export interface BlogState {
```
Comment (dev1B) [inline src/types/blogTypes.tsx:15]: weird

File: src/pages/SignIn/SignIn.tsx [hunk 15]
```diff
@@ -5,7 +5,7 @@ import { useForm, Controller } from 'react-hook-form';
 import GoogleLogin from 'react-google-login';
 import FacebookLogin from 'react-facebook-login';
 import { GoogleOutlined } from '@ant-design/icons';
-import FormErrorMessage from '../../components/FormErrorMessage/FormErrorMessage';
+import FormErrorMessage from '../../components/FormErrorMessage/FormErrorMassage';
```
Comment (dev1B) [inline src/pages/SignIn/SignIn.tsx:8]: Massage ?

File: src/pages/BlogPost/BlogPostContainer.tsx [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+  onCommentAdded: (value: string, parentId: number | null) => void;
+}
+const postId = 1;
+function mapStateToProps(state: InitialState) {
+  const posts = BlogSelectors.postsSelector(state);
+  return {
+    post: posts.find((post: Post) => post.postId === postId),
+  };
+}
+function mapDispatchToProps(dispatch: typeof store.dispatch) {
```
Comment (dev1B) [inline src/pages/BlogPost/BlogPostContainer.tsx:22]: const actions = {
  addComment: BlogActions.addComment,
  };

**Review by dev1B — CHANGES_REQUESTED** [review]

File: src/reducers/blog.tsx (hunk 9 above)
Comment (olehrdk) [reply src/reducers/blog.tsx]: nope, no-plusplus eslint rule

File: src/components/AddCommentForm/AddCommentForm.tsx [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+  const addComment = (data: FormInput) => onSubmit(data.commentText);
+
+  const textAreaRef = useRef<any>(null);
+
+  if (focus) focusOnTextArea(textAreaRef);
+
+  let disabled: boolean = true;
+  const textAreaValue = watch('commentText');
+
+  if (textAreaValue && textAreaValue.length !== 0) disabled = false;
```
Comment (dev1B) [inline src/components/AddCommentForm/AddCommentForm.tsx:30]: textAreaValue && textAreaValue.length => textAreaValue?.length

File: src/components/AddCommentForm/AddCommentForm.tsx [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+const AddCommentForm = ({ onSubmit, focus }: AddCommentFormProps) => {
+  const { register, handleSubmit, watch, reset } = useForm<FormInput>();
+
+  const addComment = (data: FormInput) => onSubmit(data.commentText);
+
+  const textAreaRef = useRef<any>(null);
+
+  if (focus) focusOnTextArea(textAreaRef);
+
+  let disabled: boolean = true;
```
Comment (dev1B) [inline src/components/AddCommentForm/AddCommentForm.tsx:27]: https://react-hook-form.com/api/useformstate#useFormRef

**Review by dev1B — CHANGES_REQUESTED** [review]

**Comment by dev1B** [comment]
Need to complete and reopen. No updates for a long time

**Pull request closed without merging**